linera-core/src/client/validator_manager/manager.rs

```diff
@@ -0,0 +1,1268 @@
+        )
+    });
+
+    /// Counter of successful requests per validator
+    pub static VALIDATOR_REQUEST_SUCCESS: LazyLock<IntCounterVec> = LazyLock::new(|| {
+        register_int_counter_vec(
+            "validator_manager_request_success",
+            "Number of successful requests to each validator",
+            &["validator"],
+        )
+    });
+
+    /// Counter for requests that were deduplicated (joined an in-flight request)
+    pub static REQUEST_CACHE_DEDUPLICATION: LazyLock<IntCounter> = LazyLock::new(|| {
+        register_int_counter(
+            "validator_manager_request_deduplication_total",
+            "Number of requests that were deduplicated by joining an in-flight request",
+        )
+    });
+
+    /// Counter for requests that were served from cache
+    pub static REQUEST_CACHE_HIT: LazyLock<IntCounter> = LazyLock::new(|| {
+        register_int_counter(
+            "validator_manager_request_cache_hit_total",
+            "Number of requests that were served from cache",
+        )
+    });
+}
+
+/// Manages a pool of validator nodes with intelligent load balancing and performance tracking.
+///
+/// The `ValidatorManager` maintains performance metrics for each validator node using
+/// Exponential Moving Averages (EMA) and uses these metrics to make intelligent routing
+/// decisions. It prevents node overload through request capacity limits and automatically
+/// retries failed requests on alternative nodes.
+///
+/// # Examples
+///
+/// ```ignore
+/// // Create with default configuration (balanced scoring)
+/// let manager = ValidatorManager::new(validator_nodes);
+///
+/// // Create with custom configuration prioritizing low latency
+/// let latency_weights = ScoringWeights {
+///     latency: 0.6,
+///     success: 0.3,
+///     load: 0.1,
+/// };
+/// let manager = ValidatorManager::with_config(
+///     validator_nodes,
+///     15,                      // max 15 concurrent requests per node
+///     latency_weights,         // custom scoring weights
+///     0.2,                     // higher alpha for faster adaptation
+///     3000.0,                  // max expected latency (3 seconds)
+///     Duration::from_secs(60), // 60 second cache TTL
+///     200,                     // cache up to 200 entries
+/// );
+/// ```
+#[derive(Debug, Clone)]
+pub struct ValidatorManager<Env: Environment> {
+    /// Thread-safe map of validator nodes indexed by their public keys.
+    /// Each node is wrapped with EMA-based performance tracking information.
+    nodes: Arc<tokio::sync::RwLock<BTreeMap<ValidatorPublicKey, NodeInfo<Env>>>>,
+    /// Maximum number of concurrent requests allowed per node.
+    /// Prevents overwhelming individual validators with too many parallel requests.
+    max_requests_per_node: usize,
+    /// Default scoring weights applied to new nodes.
+    default_weights: ScoringWeights,
+    /// Default EMA smoothing factor for new nodes.
+    default_alpha: f64,
```

Comment:
(We are always using the same smoothing factor for all nodes in any given `ValidatorManager`, are we? I'd omit the `default_` here, then, because it suggests that it can be changed for individual nodes. Same for the weights and latency.)